### PR TITLE
Periodically sync with Upstream

### LINKED ISSUE #6: Periodically sync with Upstream
- We should periodically (on Monday & Thursday each week) check if there are new commits on the upstream `XKCP/XKCP/master` branch.
- If there are new commits upstream, we should create a PR to merge them from `XKCP/XKCP/master` into `StirlingLabs/XKCP/master`.  
- If there is already a PR from a previous update that hasn't already been merged, that PR should just be updated and no new PR created.

The [Create PR from upstream repos for private forks](https://GitHub.com/marketplace/actions/create-pr-from-upstream-repos-for-private-forks)	Action looks like it does exactly this, although it has 0 stars.  The [Create Sync PR](https://github.com/marketplace/actions/create-sync-pr) Action first creates a tracking branch and then creates a PR from that, which leaves a branch that we have to delete.

### PR DESCRIPTION
**Pull request by ilkerhalil**

### CODE REVIEW

File: .github/workflows/fork-sync.yml [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        target_branch: stirling-labs
+        git_pull_args: '--unshallow --allow-unrelated-histories'
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+    - name: Create new commit
+      if: steps.sync.outputs.has_new_commits
+      run: |
+        git config --global user.email "daemon@stirlinglabs.com"
+        git config --global user.name "StirlingLabsDaemon"
+        git add -A
+        git commit -m "XKCP/XKCP main branch has changes"
```
Comment (TYoungSL) [inline .github/workflows/fork-sync.yml]: what are we committing? the previous action should do commits?

File: .github/workflows/fork-sync.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      uses: actions/checkout@v2
+      with:
+        ref: stirling-labs
+    - name: Pull upstream changes
+      id: sync
+      uses: aormsby/Fork-Sync-With-Upstream-action@v2.3
+      with:
+        upstream_repository: XKCP/XKCP
+        upstream_branch: master
+        target_branch: stirling-labs
```
Comment (TYoungSL) [inline .github/workflows/fork-sync.yml]:
```suggestion
        target_branch: master
```

File: .github/workflows/fork-sync.yml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        git_pull_args: '--unshallow --allow-unrelated-histories'
+        github_token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (TYoungSL) [inline .github/workflows/fork-sync.yml]:
```suggestion
```

File: .github/workflows/fork-sync.yml [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+      with:
+        ref: stirling-labs
+    - name: Pull upstream changes
+      id: sync
+      uses: aormsby/Fork-Sync-With-Upstream-action@v2.3
+      with:
+        upstream_repository: XKCP/XKCP
+        upstream_branch: master
+        target_branch: stirling-labs
+        git_pull_args: '--unshallow --allow-unrelated-histories'
```
Comment (TYoungSL) [inline .github/workflows/fork-sync.yml]:
```suggestion
        git_pull_args: '--unshallow'
```
I don't think we want to merge unrelated histories... They should be linear to merge with respect to history.

File: .github/workflows/fork-sync.yml [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+      if: steps.sync.outputs.has_new_commits
+      uses: peter-evans/create-pull-request@v3
+      with:
+        token:  ${{ github.token }}
+        title: 'XKCP/XKCP main branch has changes'
+        branch: check-xkcp-upstream
+        body: |
+          - Please check changes upstream XKCP master branch.
+        labels: |
+          automated-pr
```
Comment (TYoungSL) [inline .github/workflows/fork-sync.yml]:
```suggestion
          automated-pr
          xkcp
```

File: .github/workflows/fork-sync.yml [hunk 6]
```diff
@@ -0,0 +1,40 @@
+on:
+  schedule:
+    - cron:  '0 7 * * 1,4'
+  workflow_dispatch:
```
Comment (TYoungSL) [inline .github/workflows/fork-sync.yml:4]: Add inputs to match version specified in the StirlingLabs/XKCP workflow.

Example input:
```json
{
    "event_type": "New XKCP Release",
    "client_payload": { "tag": "v21.07.0" }
}
```

Reference: [GitHub Actions: Manual triggers with workflow_dispatch [2020/07/06]](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/)

File: .github/workflows/fork-sync.yml [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        git commit -m "XKCP/XKCP main branch has changes"
+    - name: Create Pull Request
+      if: steps.sync.outputs.has_new_commits
+      uses: peter-evans/create-pull-request@v3
+      with:
+        token:  ${{ github.token }}
+        title: 'XKCP/XKCP main branch has changes'
+        branch: check-xkcp-upstream
+        body: |
+          - Please check changes upstream XKCP master branch.
```
Comment (TYoungSL) [inline .github/workflows/fork-sync.yml]: Decorate title, branch and body with version tag from workflow dispatch event input. (e.g. for branch, check-xkcp-upstream-[version-tag-goes-here])

**Review by TYoungSL — CHANGES_REQUESTED** [review]

File: .github/workflows/fork-sync.yml (hunk 1 above)
Comment (ilkerhalil) [reply .github/workflows/fork-sync.yml]: You're right, I missed a new commit here.

File: .github/workflows/fork-sync.yml [hunk 8]
```diff
@@ -0,0 +1,39 @@
+on:
+  schedule:
+    - cron:  '0 7 * * 1,4'
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: 'Tag'
+        required: true
+        default: 'v21.07.0'
```
Comment (TYoungSL) [inline .github/workflows/fork-sync.yml]:
```suggestion
```
We probably want it to fail if the tag isn't specified in the dispatch request.

**Review by TYoungSL — APPROVED** [review]

**Comment by TYoungSL** [comment]
Tying to #6

**Comment by TYoungSL** [comment]
Followed up with https://github.com/StirlingLabs/XKCP/commit/eaf322d373b2ca10568c88c83ec6ff1d06a0c37e

**Pull request merged**